### PR TITLE
Start Firefox earlier on Live media

### PR DESCRIPTION
**Pull request by jkonecny12**
The execution time between Gnome Initial Setup and Anaconda is quite big
we need to reduce it by moving FF execution sooner before Anaconda is
executed.

Time to show Firefox window based on my measurements on VM with 2GB RAM
and 4vCPUs:
Original execution time: 13 seconds
With this commit: less than 6 seconds

If we want to reduce this time even more we need to do that in Firefox
or Gnome Shell (Gnome Initial Setup environment) somehow.

**TODO:**
- [x] Tests
- [ ] Add JSX support so we don't have just white screen
- ~[ ] Add event loop also to original `webui-desktop` script execution code?~
  - ~This part is a question because I think we could have an issue without any event loop running but so far there are no complains about that...~
  - Let's do that in separate PR so we don't complicate things here. The code won't be used on Live anyway.
- [ ] Make the code more robust to not crash on unexpected inputs

*Resolves: rhbz#2236438*

### CODE REVIEW

**Comment by VladimirSlavik** [comment]
Most parts tentatively LGTM, except for opening the viewer again when the webui class "starts". I think it should be possible to have the page just ... do something ... and run the starter script again only outside of live?

File: pyanaconda/ui/webui/__init__.py [hunk 1]
```diff
@@ -90,25 +93,52 @@ def run(self):
...
-        # FIXME: looks like "type" should not be used and _is_live_os is private ?
         if conf.system.provides_liveuser:
-            profile_name = FIREFOX_THEME_LIVE
+            self._watch_webui_on_live()
         else:
-            profile_name = FIREFOX_THEME_DEFAULT
+            self._run_webui()
+
+    def _run_webui(self):
+        # FIXME: looks like "type" should not be used and _is_live_os is private ?
```
Comment (rvykydal) [inline pyanaconda/ui/webui/__init__.py:97]: I wonder if this comment is still relevant.

**Comment by rvykydal** [comment]
> Most parts tentatively LGTM, except for opening the viewer again when the webui class "starts". I think it should be possible to have the page just ... do something ... and run the starter script again only outside of live?

Hm, do you mean opening FF even when user starts installation, like in liveinst script? I wonder if it would get us some gain. Maybe it could be the next step. But I didn't think much about advantages or caveats of running the script outside anaconda.

**Comment by VladimirSlavik** [comment]
> Hm, do you mean opening FF even when user starts installation, like in liveinst script?

That's what this PR is doing...

My comment was, the ui class still starts the browser, so it is opened twice. (Or is it not?)

**Comment by rvykydal** [comment]
> > Hm, do you mean opening FF even when user starts installation, like in liveinst script?
> 
> That's what this PR is doing...
> 
> My comment was, now the liveinst script starts the browser, then the ui class still starts the browser too, so it is opened twice.

Oh, valid point. Sorry for the noise I somehow overlooked the liveinst script part.

**Comment by jkonecny12** [comment]
> > Hm, do you mean opening FF even when user starts installation, like in liveinst script?
> 
> That's what this PR is doing...
> 
> My comment was, now the liveinst script starts the browser, then the ui class still starts the browser too, so it is opened twice.

It is not opening FF twice. There are just two paths to be able to split stuff and avoid doing all the changes at once.

One path:
On Live, you are starting FF with the `liveinst` script and then in the `run` method of the `CockpitUserInterface` class will go to the `_watch_webui_on_live` which will just watch the pid stored in `liveinst` script. Firefox is not opened in Anaconda at all.

Second path:
*Not* on Live, it's working as it was before because `conf.system.provides_liveuser` (should change honestly) is false so just running Firefox as before.


Ideally we want to change the Second path in the future commit and move FF execution to a script opened by our service file.

**Review by KKoukiou — COMMENTED** [review]
I can't make this work.
Last I tried it crashed with the following: 
Screen Shot 2023-09-12 at 08 29 22


I tried using this ISO https://github.com/cockpit-project/bots/pull/5241 and then applying updates.img by starting liveinst --graphical --updates=http://10.0.2.2:8002/updates.img as I explained yesterday

File: pyanaconda/ui/webui/__init__.py [hunk 2]
```diff
@@ -90,25 +88,52 @@ def run(self):
...
+        Anaconda main process will stop.
+        """
+        log.debug("web-ui: watching webui-desktop pid on live environment")
+        self._main_loop = create_main_loop()
+        pid = -1
+
+        with open("/run/anaconda/webui_script.pid", "tr") as f:
+            pid = int(f.readline().strip())
+
+        PidWatcher.watch_process(pid, self._webui_desktop_closed)
```
Comment (rvykydal) [inline pyanaconda/ui/webui/__init__.py:124]: I think this should be
```
PidWatcher().watch_process(pid, self._webui_desktop_closed)
```

Comment (jkonecny12) [reply pyanaconda/ui/webui/__init__.py]: Good point. Fixed.

Comment (jkonecny12) [reply pyanaconda/ui/webui/__init__.py]: This happens if you do changes to the code during creation of tests and not testing it in production afterwards.

**Comment by VladimirSlavik** [comment]
> It is not opening FF twice. There are just two paths to be able to split stuff and avoid doing all the changes at once.

Ha, true. Then this should be fine for master.

**Comment by jkonecny12** [comment]
Closing in favor of https://github.com/rhinstaller/anaconda/pull/5181

**Pull request closed without merging**